build.sh

```diff
@@ -0,0 +1,149 @@
+#!/bin/bash
+ROOT_DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )
+BUILD_DIR="$ROOT_DIR/build"
+RELEASE_DIR="$ROOT_DIR/_RELEASE"
+
+# Removes existing libraries
+rm -f "$RELEASE_DIR/sfml-audio-2.dll"
+rm -f "$RELEASE_DIR/sfml-graphics-2.dll"
+rm -f "$RELEASE_DIR/sfml-network-2.dll"
+rm -f "$RELEASE_DIR/sfml-system-2.dll"
+rm -f "$RELEASE_DIR/sfml-window-2.dll"
+rm -f "$RELEASE_DIR/openal32.dll"
+rm -f "$RELEASE_DIR/libzlib1.dll"
+rm -f "$RELEASE_DIR/OHWorkshopUploader.exe"
+rm -f "$RELEASE_DIR/SSVOpenHexagon.exe"
+
+# Argument variable declarations
+REGENERATE_CMAKE=0
+RUN=0
+DEBUG=0
+VALGRIND=0
+MAKE_THREAD_COUNT="-j1"
+
+POSITIONAL=()
+
+# Gets all arguments specified
+while [[ $# -gt 0 ]]; do
+    arg="$1"
+    case $arg in
+        -h|--help)
+            echo "Usage:                     ./build.sh ..."
+            echo "Arguments:"
+            echo "    -r, --run              | Runs the game after build completion."
+            echo "    -d, --debug            | Runs and debugs the game after build completion."
+            echo "    -v, --valgrind         | Valgrinds the game after build completion. (Disabled for MinGW.)"
+            echo "    -g, --regenerate-cmake | Regenerates CMake files in build folder to match current OS."
+            echo "                              (This is automatic, but can be done manually if needed.)"
+	        echo "    -jN                    ! Executes the 'make' command using N threads. Default is 1."
+            echo "    -h, --help             | Displays this help."
+            echo ""
+            exit 1
+            ;;
+
+        -r|--run)
+            RUN=1
+            shift
+            ;;
+
+        -d|--debug)
+            DEBUG=1
+            shift
+            ;;
+
+        -v|--valgrind)
+            VALGRIND=1
+            shift
+            ;;
+        
+        -g|--regenerate-cmake)
+            REGENERATE_CMAKE=1
+            shift
+            ;;
+
+        -j*)
+            MAKE_THREAD_COUNT=$arg
+            shift
+            ;;
+
+        *)
+            POSITIONAL+=("$1")
+            shift
+            ;;
+    esac
+done
+
+set -- "${POSITIONAL[@]}"
+
+# Enters into the build directory
+mkdir -p $BUILD_DIR
+cd $BUILD_DIR
+
+# Determines if the CMake files need to be regenerated to match the current OS
+if [[ -f "cmake-ostype" ]]; then
+    read -r CMAKE_OSTYPE < "cmake-ostype"
+    if [[ $CMAKE_OSTYPE != $OSTYPE ]]; then
+        REGENERATE_CMAKE=1
+    fi
+else
+    REGENERATE_CMAKE=1
+fi
+
+case $OSTYPE in
+    # MinGW / Windows...
+    "msys")
+        # Disables the valgrind argument.
+        if [[ $VALGRIND -eq 1 ]]; then
+            echo "Cannot valgrind in MinGW. Disabling this argument."
+            VALGRIND=0
+        fi
+
+        if [[ $REGENERATE_CMAKE -eq 1 ]]; then
+            echo "Regenerating CMake Files for MinGW/Windows..."
+            rm -rf ./**
```

Comment:
This is scary, can we have a little bit more path here? Maybe something that includes `$BUILD_DIR` so that if something goes wrong the script doesn't delete all the user's files?